### PR TITLE
Notification manager

### PR DESCRIPTION
**Pull request by nickvergessen**
### To do list:
- [x] Check back with desktop client :speech_balloon:
- [ ] Check back with mobile clients :speech_balloon:
- [x] Check back with Tasks developers: @Henni @raimund-schluessler :speech_balloon:
- [x] Add tests :construction: 
- [x] Write an App for it :seedling:

### CODE REVIEW

**Comment by karlitschek** [comment]
sweet. :-)

**Comment by raimund-schluessler** [comment]
Wow, this was fast. :smiley: Looking forward to test it with Tasks.
Are there plans who could maintain the necessary notifications app?

**Comment by nickvergessen** [comment]
> Are there plans who could maintain the necessary notifications app?

Since this is kind of fundamental and will also make it into the desktop and mobile clients, I guess we will take care of it.

**Comment by karlitschek** [comment]
so ready to merge? Would be good to have this today or tomorrow.

**Comment by nickvergessen** [comment]
I will write some tests for the manager now, but should be ready later today

**Comment by nickvergessen** [comment]
Should be ready now, please review @tomneedham @raimund-schluessler @Henni 

To test it:

1. install the notification app (master) and check out this branch go to `/index.php/apps/notifications` which should be empty (`[]`).

2. Create a remote share with yourself and then visit `/index.php/apps/notifications` and it should not be empty. 

3. Then accept the share and `/index.php/apps/notifications` should be empty again

**Comment by karlitschek** [comment]
:+1:

**Comment by raimund-schluessler** [comment]
Hm, I seem to do something wrong. I checked everything out and created a remote share (```admin@server/owncloud```), but nothing happens under ```/index.php/apps/notifications```. However, the remote share doesn't show up in my files either? Is this right or is something wrong with remote sharing (I didn't use it yet)?

**Comment by raimund-schluessler** [comment]
Nevermind, my login name was ```Admin@server/owncloud``` obviously (Capitalization matters ;). Everything works as advertised now, nice job.

:+1:

**Comment by raimund-schluessler** [comment]
@nickvergessen Will this make it into 8.2? Looks like everything works as expected, so it should be fine, shouldn't it?

**Comment by DeepDiver1975** [comment]
If the API is meant to be experimental for now: skip ocp and only provide the private interfaces and classes.

**Comment by karlitschek** [comment]
O.K. Let's do it as @DeepDiver1975 suggested as experimental API for now. I also added it to the daily builds.
@nickvergessen Is https://github.com/owncloud/notifications/pull/10 ready for merge?

**Comment by nickvergessen** [comment]
@karlitschek I'm not a UI expert. I didn't get it working, but IIRC @tomneedham is afk. So someone else needs to have a look.

Will fix the OCP vs OC thing tomorrow then.

**Comment by raimund-schluessler** [comment]
owncloud/notifications#10 is surely an improvement and first step compared to not having any UI at all, but it definitively needs some polishing ;)

**Comment by karlitschek** [comment]
@tomneedham is offline at the moment.
@jancborchardt @PVince81 What do you think?

**Comment by nickvergessen** [comment]
PVinc is on vacation this week

**Comment by karlitschek** [comment]
@nickvergessen true @jancborchardt who do you think can have a look? or is the current state fine already?

**Comment by nickvergessen** [comment]
> If the API is meant to be experimental for now: skip ocp and only provide the private interfaces and classes.

Done

Also rebased to fix the "merge conflict"

**Comment by nickvergessen** [comment]
@karlitschek well basically this PR here in core can be merged without the UI PR in the notifications app.
It just allows other apps to send notifications. If the notifications app is not enabled they will just never show up anywhere. So that shouldn't be a problem.

**Comment by jancborchardt** [comment]
I’ll work on the UI in #10. Go ahead and merge this if it’s separate.

File: lib/private/notification/action.php [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	/**
+	 * Constructor
+	 */
+	public function __construct() {
+		$this->label = '';
+		$this->labelParsed = '';
+		$this->link = '';
+		$this->requestType = '';
+		$this->icon = '';
+	}
```
Comment (LukasReschke) [inline lib/private/notification/action.php]: What is the use-case for not having this directly in the member variables defined?

File: lib/private/notification/notification.php [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+		$this->subjectParameters = [];
+		$this->subjectParsed = '';
+		$this->message = '';
+		$this->messageParameters = [];
+		$this->messageParsed = '';
+		$this->link = '';
+		$this->icon = '';
+		$this->actions = [];
+		$this->actionsParsed = [];
+	}
```
Comment (LukasReschke) [inline lib/private/notification/notification.php]: What is the use-case for not having this directly in the member variables defined?

File: lib/private/notification/manager.php [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	protected $appsClosures;
+
+	/** @var \Closure */
+	protected $notifiersClosures;
+
+	public function __construct() {
+		$this->apps = [];
+		$this->notifiers = [];
+		$this->appsClosures = [];
+		$this->notifiersClosures = [];
```
Comment (LukasReschke) [inline lib/private/notification/manager.php]: What is the use-case for not having this directly in the member variables defined?

File: lib/private/notification/manager.php [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+		if (!($notification instanceof INotification) || !$notification->isValidParsed()) {
+			throw new \InvalidArgumentException('The given notification has not been handled');
+		}
+
+		return $notification;
+	}
+
+	/**
+	 * @param INotification $notification
+	 * @return null
```
Comment (LukasReschke) [inline lib/private/notification/manager.php]: void?

File: lib/private/notification/notification.php [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+	 */
+	public function setMessage($message, array $parameters = []) {
+		if (!is_string($message) || $message === '' || isset($message[64])) {
+			throw new \InvalidArgumentException('The given message is invalid');
+		}
+		$this->message = $message;
+
+		if (!is_array($parameters)) {
+			throw new \InvalidArgumentException('The given message parameters are invalid');
+		}
```
Comment (LukasReschke) [inline lib/private/notification/notification.php]: Not required as type is enforced before?

File: lib/private/notification/notification.php [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+	 * @throws \InvalidArgumentException if the subject or parameters are invalid
+	 * @since 8.2.0
+	 */
+	public function setSubject($subject, array $parameters = []) {
+		if (!is_string($subject) || $subject === '' || isset($subject[64])) {
+			throw new \InvalidArgumentException('The given subject is invalid');
+		}
+		$this->subject = $subject;
+
+		if (!is_array($parameters)) {
```
Comment (LukasReschke) [inline lib/private/notification/notification.php]: Not required since type is enforced before?

File: lib/private/notification/action.php (hunk 1 above)
Comment (nickvergessen) [inline lib/private/notification/action.php]: personal preference ;)

File: lib/private/notification/notification.php (hunk 5 above)
Comment (nickvergessen) [inline lib/private/notification/notification.php]: Yeah I added the typehint later and didn't remove this, as it doesn't hurt

**Comment by jancborchardt** [comment]
Well, at least with this pull request for me the notifications app doesn’t break master. ;)

**Comment by LukasReschke** [comment]
Let's get this in.

**Pull request merged**